### DISCUSSION TITLE
How to group "major" and "minor" for specific updates when using :separateMajorReleases

### DISCUSSION THREAD

**Discussion by jpgoelz** [renovatebot/renovate#25831] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

37.59.8

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

We _usually_ want to separate minor and major releases, which is why we are using `:separateMajorReleases` and `:separatePatchReleases`.

However, in the complicated world of Angular updates, we need to group some major updates with certain minor updates.

For example:
There is an update from Angular 16 to 17 at the moment.
However, `@angular-devkit/architect`, which is part of Angular, holds the major version in the minor field for some reason: Update from 0.1602.10 to 0.1700.1 instead of 16.2.10 to 17.0.1.
The same is true for zone.js, which is part of Angular as well. Version 0.13.x was for Angular 16 while 0.14.x is for Angular 17.

Angular is very strict about the needed versions, and the latter always leads to the Renovate merge request not being able to `npm install`:

![image](https://github.com/renovatebot/renovate/assets/32487587/8a443b9b-9648-4c22-aaaa-4ad152720241)

So tl;dr: The wanted end result is a package rule (or set thereof) that makes it possible to combine the needed minor upgrades into the same branch as the major upgrade while still separating major, minor, and patch releases for all other dependencies.

### What you tried so far.

We currently have tried these rules:

```json5
  packageRules: [
    {
      groupName: "Angular Major",
      groupSlug: "angular-major",
      matchUpdateTypes: ["major"],
      matchSourceUrls: [
        "https://github.com/angular/angular",
        "https://github.com/angular/angular-cli",
      ],
    },
    {
      groupName: "Angular Major",
      matchUpdateTypes: ["minor"],
      matchPackageNames: ["zone.js", "@angular-devkit/architect"],
    },
  }
```

However, because we extend `:separateMajorReleases` and `:separatePatchReleases`, we get separate merge requests:

![image](https://github.com/renovatebot/renovate/assets/32487587/d1db988f-a81a-482c-b7c1-1d3836a0b1e2)

**Comment by rarkins** [comment]
> We _usually_ want to separate minor and major releases, which is why we are using `:separateMajorReleases` and `:separatePatchReleases`.

Neither of these settings are relevant to "separate minor and major releases". The relevant settings is separateMajorMinor, which defaults to true. You should remove both these settings if that's your only reason, because it will make everything else you're asking about more complicated

**Reply by jpgoelz** [reply]
Let me clarify:
We use these settings, because we want to have separate merge requests for patch, minor, and major versions.
For example: A dependency may be at v1.2.3. Our build breaks for v1.3.0, but we do not want to miss 1.2.4. One of those cases is Spring Boot. We only got updates for v3, but are at v2.7 and did not get the patch or minor updates separately.

Also this confuses me:
> Neither of these settings are relevant to "separate minor and major releases". The relevant settings is separateMajorMinor, [...]

According to the docs, `:separateMajorReleases` is exactly `separateMajorMinor: true`. Or do I misunderstand this?
![image](https://github.com/renovatebot/renovate/assets/32487587/874a669a-3348-489d-a1b5-665573eb1016)

**Reply by rarkins** [reply]
Sorry you're right - that name confused me (thought it was separateMultipleMajor=true).

**Comment by rarkins** [comment]
What are the current branch names for the two updates you want to combine?

**Reply by jpgoelz** [reply]
The names are `renovate/major-angular-major` for "Update Angular Major to v17 (major)" and `renovate/angular-major` for "Update Angular Major (minor)".

**Reply by rarkins** [reply, marked as answer]
Try adding `"groupSlug": "major-angular-major"` to the second packageRule

**Reply by jpgoelz** [reply]
That worked, thanks! 

The dependencies are now grouped as expected. The Title now reads "Update Angular Major (minor)" but I guess I can fix that with `prTitle`.

However, now of course TypeScript needs to also increase the version and npm won't install. Renovate and Angular just don't play that well together. 😅 But that is not relevant for this discussion.

**Comment by joloto** [comment]
I came across this thread as I had the same desire to group updates of `@angular-devkit/architect` with angular major updates.

I couldn't get the solution above to work. 

But I was able to solve it by configuring the versioning of `@angular-devkit/architect`, to treat the first 2 digits of the minor version part as major, i.e.:

```
packageRules: [
    ...
    {
      "matchPackageNames": ["@angular-devkit/architect"],
      "versioning": "regex:^[~^]?0\\.(?<major>\\d{2})(?<minor>\\d+)\\.(?<patch>\\d+)$"
    },
    {
      groupName: "Angular Major",
      groupSlug: "angular-major",
      matchUpdateTypes: ["major"],
      matchSourceUrls: [
        "https://github.com/angular/angular",
        "https://github.com/angular/angular-cli",
      ],
    },
    ...
  }
```

this way the correct corresponding version of `@angular-devkit/architect` is included with configured grouped angular updates.

**Reply by jpgoelz** [reply]
Oh that is nice. I will try it when I get a chance. Thanks!
